File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    console.log('손잡이', hand);
    const Ls = ["*", 7, 4, 1];
    const Rs = ["#", 9, 6, 3 ];
    const Ns = [0, 8, 5, 2];
    
    let lPos = "*";
    let rPos = "#";
    
    const hands =  numbers.map(num => {
        if(Ls.includes(num)){
            lPos = num;
            return "L";
        } else if(Rs.includes(num)){
            rPos = num;
            return "R";
        } else {
            // 현재 손가락의 위치에서 num까지의 거리를 구해야 함 
            console.log("숫자", num, "왼쪽", lPos, "오른쪽", rPos);
            
            const numIndex = Ns.indexOf(num);
            const leftIndex = Ls.indexOf(lPos) === -1 ? Ns.indexOf(lPos) : Ls.indexOf(lPos);
            const rightIndex = Rs.indexOf(rPos) === -1 ? Ns.indexOf(rPos) : Rs.indexOf(rPos);
            
            console.log(leftIndex, rightIndex)
            
            const leftDist = Ls.indexOf(lPos) === -1 ? Math.abs(Ns.indexOf(num) - Ns.indexOf(lPos)) : Math.abs(numIndex - leftIndex) + 1
            const rightDist = Rs.indexOf(rPos) === -1 ? Math.abs(Ns.indexOf(num) - Ns.indexOf(rPos)) : Math.abs(numIndex - rightIndex) + 1
            
            console.log(leftDist, rightDist)
            
            if(leftDist > rightDist){
                rPos = num;
                return "R"
            } else if (leftDist < rightDist){
                lPos = num;
                return "L"
            } else {
                if(hand === "left") {
                    lPos = num;
                    return "L";
                }
                else {
                    rPos = num;
                    return "R";
                }
            }
        }
    }).join("")
    
    console.log(hands)
    
    return hands;
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    print("숫자 열", numbers)
    print("손 잡이", hand)
    
    keypad_dict = {
        1:(1, 1),
        2: (1, 2),
        3: (1, 3),
        4:(2, 1),
        5: (2, 2),
        6: (2, 3),
        7:(3, 1),
        8: (3, 2),
        9: (3, 3),
        "*":(4, 1),
        0: (4, 2),
        "#": (4, 3),
    }
    
    l_keys = {1, 4, 7}
    r_keys = {3, 6, 9}
    
    cur_l = '*'
    cur_r = "#"
    
    result = ""
    
    for num in numbers:
        if num in l_keys:
            result += "L"
            cur_l = num
        elif num in r_keys:
            result += "R"
            cur_r = num
        else:
            [nh, nw] = keypad_dict[num]
            [lh, lw] = keypad_dict[cur_l]
            [rh, rw] = keypad_dict[cur_r]
            l_dist = abs(nh - lh) + abs(nw - lw)
            r_dist = abs(nh - rh) + abs(nw - rw)
            
            if l_dist < r_dist:
                result += "L"
                cur_l = num
            elif l_dist > r_dist:
                result += "R"
                cur_r = num
            else:
                if hand == "left":
                    result += "L"
                    cur_l = num
                else:
                    result += "R"
                    cur_r = num
            
    return result